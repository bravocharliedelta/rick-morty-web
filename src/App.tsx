import { Routes, Route, Navigate } from 'react-router-dom';
import {
  CacheConfigProvider,
  CharactersList,
  AuthProvider,
  Login,
  Layout,
  RequireAuth,
  NotFound,
  StylesProvider,
} from './components';

function App() {
  return (
    <AuthProvider>
      <CacheConfigProvider>
        <StylesProvider>
          <Routes>
            <Route path="/" element={<Layout />}>
              <Route path="login" element={<Login navigateTo="/characters" />} />
              <Route element={<RequireAuth />}>
                <Route path="characters" element={<CharactersList />} />
              </Route>
              <Route path="" element={<Navigate to="/login" replace />} />
              <Route path="*" element={<NotFound />} />
            </Route>
          </Routes>
        </StylesProvider>
      </CacheConfigProvider>
    </AuthProvider>
  );
}

export default App;
