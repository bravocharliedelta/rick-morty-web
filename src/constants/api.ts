const RM_API_BASE_URL = '/api';
const CHARACTERS_API = `${RM_API_BASE_URL}/characters`;

export { RM_API_BASE_URL, CHARACTERS_API };
