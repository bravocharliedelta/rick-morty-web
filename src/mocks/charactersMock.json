{
  "info": {
    "count": 826,
    "pages": 42,
    "next": "https://rickandmortyapi.com/api/character?page=2",
    "prev": null
  },
  "results": [
    {
      "id": 1,
      "name": "Rick Sanchez",
      "status": "Alive",
      "favorite": false,
      "origin": "Earth (C-137)",
      "url": "https://rickandmortyapi.com/api/location/1",
      "location": "Citadel of Ricks",
      "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg"
    },
    {
      "id": 2,
      "name": "Morty Smith",
      "status": "Alive",
      "favorite": false,
      "origin": "unknown",
      "url": "",
      "location": "Citadel of Ricks",
      "image": "https://rickandmortyapi.com/api/character/avatar/2.jpeg"
    },
    {
      "id": 3,
      "name": "Summer Smith",
      "status": "Alive",
      "favorite": false,
      "origin": "Earth (Replacement Dimension)",
      "location": "Earth (Replacement Dimension)",
      "image": "https://rickandmortyapi.com/api/character/avatar/3.jpeg"
    },
    {
      "id": 4,
      "name": "Beth Smith",
      "status": "Alive",
      "favorite": false,
      "origin": "Earth (Replacement Dimension)",
      "location": "Earth (Replacement Dimension)",
      "image": "https://rickandmortyapi.com/api/character/avatar/4.jpeg"
    },
    {
      "id": 5,
      "name": "Jerry Smith",
      "status": "Alive",
      "favorite": false,
      "origin": "Earth (Replacement Dimension)",
      "location": "Earth (Replacement Dimension)",
      "image": "https://rickandmortyapi.com/api/character/avatar/5.jpeg"
    },
    {
      "id": 6,
      "name": "Abadango Cluster Princess",
      "status": "Alive",
      "favorite": false,
      "origin": "Abadango",
      "location": "Abadango",
      "image": "https://rickandmortyapi.com/api/character/avatar/6.jpeg"
    },
    {
      "id": 7,
      "name": "Abradolf Lincler",
      "status": "unknown",
      "favorite": false,
      "origin": "Earth (Replacement Dimension)",
      "location": "Testicle Monster Dimension",
      "image": "https://rickandmortyapi.com/api/character/avatar/7.jpeg"
    },
    {
      "id": 8,
      "name": "Adjudicator Rick",
      "status": "Dead",
      "favorite": false,
      "origin": "unknown",
      "location": "Citadel of Ricks",
      "image": "https://rickandmortyapi.com/api/character/avatar/8.jpeg"
    },
    {
      "id": 9,
      "name": "Agency Director",
      "status": "Dead",
      "favorite": false,
      "origin": "Earth (Replacement Dimension)",
      "location": "Earth (Replacement Dimension)",
      "image": "https://rickandmortyapi.com/api/character/avatar/9.jpeg"
    },
    {
      "id": 10,
      "name": "Alan Rails",
      "status": "Dead",
      "favorite": false,
      "origin": "unknown",
      "location": "Worldender's lair",
      "image": "https://rickandmortyapi.com/api/character/avatar/10.jpeg"
    }
  ]
}
