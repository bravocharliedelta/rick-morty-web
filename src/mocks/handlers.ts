// src/mocks/handlers.js
// eslint-disable-next-line import/no-extraneous-dependencies
import { rest } from 'msw';
import { RM_API_BASE_URL, RM_CHARACTERS_PATH, RM_LOGIN_PATH } from '../constants/api';
import charactersMock from './charactersMock.json';

export const handlers = [
  rest.get(RM_API_BASE_URL + RM_CHARACTERS_PATH, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(charactersMock));
  }),

  rest.post(RM_API_BASE_URL + RM_LOGIN_PATH, (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        userId: '_2392',
        expirationDate: new Date(new Date().getTime() + 15 * 60 * 1000).toISOString(),
      })
    );
  }),
];
