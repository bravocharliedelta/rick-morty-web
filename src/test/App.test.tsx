import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import { rest } from 'msw';
import { server } from '../mocks/server';
import { RM_API_BASE_URL, RM_CHARACTERS_PATH } from '../constants/api';
import App from '../App';

const login = async ({ email = 'rick.sanchez@pickle.org', password = 'ImPickleRick!' } = {}) => {
  await userEvent.type(screen.getByLabelText(/email address/i), email);
  await userEvent.type(screen.getByText(/password/i), password);
  await userEvent.click(screen.getByRole('button', { name: /sign in/i }));
};

/* 
  check mock folder to preview default, fetched, mocked data
*/

describe('App:', () => {
  it('shows error message when fetching characters failed', async () => {
    server.use(
      // In this paticular test respond to the same request with a 404 response.
      rest.get(`${RM_API_BASE_URL}${RM_CHARACTERS_PATH}`, (req, res, ctx) => {
        return res(ctx.status(404), ctx.json({}));
      })
    );

    render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );

    await login();

    // redirects to /character
    await screen.findByText('There was a problem loading characters, please try again');
  });

  it('shows all happy path steps', async () => {
    const { container } = render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );

    await login();

    // redirects to /character
    await screen.findByText(/loading/i);
    await screen.findByAltText(/rick sanchez/i);
    expect(container).toMatchSnapshot();
  });

  it('validates login credentials', async () => {
    render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );

    const emailInput = screen.getByLabelText(/email address/i);
    const passInput = screen.getByLabelText(/password/i);

    // submit empty form
    await userEvent.click(screen.getByRole('button', { name: /sign in/i }));
    expect(screen.getAllByText(/field is required/i).length).toBe(2);

    await userEvent.type(emailInput, 'rick.sanchez');
    expect(screen.getByText(/invalid email/i)).toBeInTheDocument();

    await userEvent.type(emailInput, 'rick.sanchez@pickle.org');
    await userEvent.type(passInput, 'ImPickleRick!');
    expect(screen.queryAllByRole('alert').length).toBe(0);
  });
});
