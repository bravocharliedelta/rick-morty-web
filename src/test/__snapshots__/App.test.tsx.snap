// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App: shows all happy path steps 1`] = `
<div>
  <main
    class="css-1foi7al"
  >
    <section>
      <ul
        class="css-hxhbq5"
      >
        <li>
          <article
            class="css-ej8t2f"
          >
            <div
              class="css-19s40fc"
            >
              <img
                alt="Rick Sanchez"
                src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
              />
            </div>
            <div
              class="css-strwe"
            >
              <h2>
                Rick Sanchez
              </h2>
              <h4>
                Alive
              </h4>
              <h3>
                Last known location:
              </h3>
              <p>
                Citadel of Ricks
              </p>
              <h3>
                First seen in:
              </h3>
              <p>
                Earth (C-137)
              </p>
              <a
                href="/character/1"
              >
                More info
              </a>
            </div>
          </article>
        </li>
        <li>
          <article
            class="css-ej8t2f"
          >
            <div
              class="css-19s40fc"
            >
              <img
                alt="Morty Smith"
                src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
              />
            </div>
            <div
              class="css-strwe"
            >
              <h2>
                Morty Smith
              </h2>
              <h4>
                Alive
              </h4>
              <h3>
                Last known location:
              </h3>
              <p>
                Citadel of Ricks
              </p>
              <h3>
                First seen in:
              </h3>
              <p>
                unknown
              </p>
              <a
                href="/character/2"
              >
                More info
              </a>
            </div>
          </article>
        </li>
        <li>
          <article
            class="css-ej8t2f"
          >
            <div
              class="css-19s40fc"
            >
              <img
                alt="Summer Smith"
                src="https://rickandmortyapi.com/api/character/avatar/3.jpeg"
              />
            </div>
            <div
              class="css-strwe"
            >
              <h2>
                Summer Smith
              </h2>
              <h4>
                Alive
              </h4>
              <h3>
                Last known location:
              </h3>
              <p>
                Earth (Replacement Dimension)
              </p>
              <h3>
                First seen in:
              </h3>
              <p>
                Earth (Replacement Dimension)
              </p>
              <a
                href="/character/3"
              >
                More info
              </a>
            </div>
          </article>
        </li>
        <li>
          <article
            class="css-ej8t2f"
          >
            <div
              class="css-19s40fc"
            >
              <img
                alt="Beth Smith"
                src="https://rickandmortyapi.com/api/character/avatar/4.jpeg"
              />
            </div>
            <div
              class="css-strwe"
            >
              <h2>
                Beth Smith
              </h2>
              <h4>
                Alive
              </h4>
              <h3>
                Last known location:
              </h3>
              <p>
                Earth (Replacement Dimension)
              </p>
              <h3>
                First seen in:
              </h3>
              <p>
                Earth (Replacement Dimension)
              </p>
              <a
                href="/character/4"
              >
                More info
              </a>
            </div>
          </article>
        </li>
        <li>
          <article
            class="css-ej8t2f"
          >
            <div
              class="css-19s40fc"
            >
              <img
                alt="Jerry Smith"
                src="https://rickandmortyapi.com/api/character/avatar/5.jpeg"
              />
            </div>
            <div
              class="css-strwe"
            >
              <h2>
                Jerry Smith
              </h2>
              <h4>
                Alive
              </h4>
              <h3>
                Last known location:
              </h3>
              <p>
                Earth (Replacement Dimension)
              </p>
              <h3>
                First seen in:
              </h3>
              <p>
                Earth (Replacement Dimension)
              </p>
              <a
                href="/character/5"
              >
                More info
              </a>
            </div>
          </article>
        </li>
        <li>
          <article
            class="css-ej8t2f"
          >
            <div
              class="css-19s40fc"
            >
              <img
                alt="Abadango Cluster Princess"
                src="https://rickandmortyapi.com/api/character/avatar/6.jpeg"
              />
            </div>
            <div
              class="css-strwe"
            >
              <h2>
                Abadango Cluster Princess
              </h2>
              <h4>
                Alive
              </h4>
              <h3>
                Last known location:
              </h3>
              <p>
                Abadango
              </p>
              <h3>
                First seen in:
              </h3>
              <p>
                Abadango
              </p>
              <a
                href="/character/6"
              >
                More info
              </a>
            </div>
          </article>
        </li>
        <li>
          <article
            class="css-ej8t2f"
          >
            <div
              class="css-19s40fc"
            >
              <img
                alt="Abradolf Lincler"
                src="https://rickandmortyapi.com/api/character/avatar/7.jpeg"
              />
            </div>
            <div
              class="css-strwe"
            >
              <h2>
                Abradolf Lincler
              </h2>
              <h4>
                unknown
              </h4>
              <h3>
                Last known location:
              </h3>
              <p>
                Testicle Monster Dimension
              </p>
              <h3>
                First seen in:
              </h3>
              <p>
                Earth (Replacement Dimension)
              </p>
              <a
                href="/character/7"
              >
                More info
              </a>
            </div>
          </article>
        </li>
        <li>
          <article
            class="css-ej8t2f"
          >
            <div
              class="css-19s40fc"
            >
              <img
                alt="Adjudicator Rick"
                src="https://rickandmortyapi.com/api/character/avatar/8.jpeg"
              />
            </div>
            <div
              class="css-strwe"
            >
              <h2>
                Adjudicator Rick
              </h2>
              <h4>
                Dead
              </h4>
              <h3>
                Last known location:
              </h3>
              <p>
                Citadel of Ricks
              </p>
              <h3>
                First seen in:
              </h3>
              <p>
                unknown
              </p>
              <a
                href="/character/8"
              >
                More info
              </a>
            </div>
          </article>
        </li>
        <li>
          <article
            class="css-ej8t2f"
          >
            <div
              class="css-19s40fc"
            >
              <img
                alt="Agency Director"
                src="https://rickandmortyapi.com/api/character/avatar/9.jpeg"
              />
            </div>
            <div
              class="css-strwe"
            >
              <h2>
                Agency Director
              </h2>
              <h4>
                Dead
              </h4>
              <h3>
                Last known location:
              </h3>
              <p>
                Earth (Replacement Dimension)
              </p>
              <h3>
                First seen in:
              </h3>
              <p>
                Earth (Replacement Dimension)
              </p>
              <a
                href="/character/9"
              >
                More info
              </a>
            </div>
          </article>
        </li>
        <li>
          <article
            class="css-ej8t2f"
          >
            <div
              class="css-19s40fc"
            >
              <img
                alt="Alan Rails"
                src="https://rickandmortyapi.com/api/character/avatar/10.jpeg"
              />
            </div>
            <div
              class="css-strwe"
            >
              <h2>
                Alan Rails
              </h2>
              <h4>
                Dead
              </h4>
              <h3>
                Last known location:
              </h3>
              <p>
                Worldender's lair
              </p>
              <h3>
                First seen in:
              </h3>
              <p>
                unknown
              </p>
              <a
                href="/character/10"
              >
                More info
              </a>
            </div>
          </article>
        </li>
      </ul>
    </section>
  </main>
</div>
`;
